### PR TITLE
Report affected rows for DML statements

### LINKED ISSUE #707: Include "number of rows affected" in the data captured for SQL DML commands
Implementing https://github.com/elastic/apm/issues/112.

Add instrumentation to the `int executeUpdate()` APIs, using the return value for this. 
If user uses the non-DML `execute` APIs- this count will not be populated.

### PR DESCRIPTION
**Pull request by SylvainJuge**
Fixes elastic/apm-agent-java#707

Reports the number of rows affected by a DML statement (INSERT,DELETE,UPDATE) in database span.

### Checklist

- [x] Implement code
- [x] Add tests
- [x] ~~Update documentation~~
- [x] Update [CHANGELOG.md](CHANGELOG.md)
- [x] ~~Update [supported-technologies.asciidoc](docs/supported-technologies.asciidoc)~~
- [x] ~~Added an API method or config option? Document in which version this will be introduced.~~

### CODE REVIEW

File: apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/StatementInstrumentation.java [hunk 1]
```diff
@@ -119,15 +184,40 @@ public static Span onBeforeExecute(@Advice.This Statement statement, @Advice.Arg
...
+                        int[] array = (int[]) returnValue;
+                        for (int i = 0; i < array.length; i++) {
+                            affectedCount += array[i];
+                        }
+                    } else if ( returnValue instanceof long[]){
+                        long[] array = (long[]) returnValue;
+                        for (int i = 0; i < array.length; i++) {
+                            affectedCount += array[i];
+                        }
+                    }
```
Comment (felixbarny) [inline apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/StatementInstrumentation.java]: Extract public static method, annotate with `@VisibleForAdvice` and reuse in `ExecuteWithoutQueryInstrumentation`. As only Java core types are used, there are no classloader issues.

File: apm-agent-plugins/apm-jdbc-plugin/src/test/java/co/elastic/apm/agent/jdbc/AbstractJdbcInstrumentationTest.java [hunk 2]
```diff
@@ -150,17 +154,34 @@ private void testBatchPreparedStatement() throws SQLException {
...
+    private void testMultipleRowsModifiedStatement() throws SQLException {
+        String insert = "INSERT INTO ELASTIC_APM (FOO, BAR) VALUES (3, 'TEST')";
+        Statement statement = connection.createStatement();
+        statement.execute(insert);
+
+        assertThat(reporter.getSpans()).hasSize(1);
+        Db db = reporter.getFirstSpan().getContext().getDb();
+        assertThat(db.getStatement()).isEqualTo(insert);
+        assertThat(db.getAffectedRowsCount())
+            .isEqualTo(statement.getUpdateCount())
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-jdbc-plugin/src/test/java/co/elastic/apm/agent/jdbc/AbstractJdbcInstrumentationTest.java]: Hmm, it seems all drivers DO support multiple invocations of `getUpdateCount` despite the javadoc. Still not sure if it is safe enough to use it...

File: apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/StatementInstrumentation.java (hunk 1 above)
Comment (eyalkoren) [reply apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/StatementInstrumentation.java]: I don't get that- those APIs return either `int` or `long`, not arrays...
When fixing, you will need to create two separate advices.

File: apm-agent-plugins/apm-jdbc-plugin/src/test/java/co/elastic/apm/agent/jdbc/AbstractJdbcInstrumentationTest.java [hunk 3]
```diff
@@ -150,17 +154,34 @@ private void testBatchPreparedStatement() throws SQLException {
...
-        assertSpanRecorded(query, false);
+        assertThat(updates[0]).isEqualTo(1);
+        assertThat(updates[1]).isEqualTo(1);
+        assertSpanRecorded(query, false, 2);
+    }
+
+    private void testMultipleRowsModifiedStatement() throws SQLException {
+        String insert = "INSERT INTO ELASTIC_APM (FOO, BAR) VALUES (3, 'TEST')";
+        Statement statement = connection.createStatement();
+        statement.execute(insert);
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-jdbc-plugin/src/test/java/co/elastic/apm/agent/jdbc/AbstractJdbcInstrumentationTest.java]: `executeUpdate` and `executeLargeUpdate` are not tested. This is what allowed the instrumentation bug.

**Comment by eyalkoren** [comment]
Also - this new value should be serialized into DB span events as part of the implementation of https://github.com/elastic/apm/issues/112

File: apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/StatementInstrumentation.java (hunk 1 above)
Comment (SylvainJuge) [reply apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/StatementInstrumentation.java]: should be fixed now

File: apm-agent-plugins/apm-jdbc-plugin/src/test/java/co/elastic/apm/agent/jdbc/AbstractJdbcInstrumentationTest.java (hunk 3 above)
Comment (SylvainJuge) [reply apm-agent-plugins/apm-jdbc-plugin/src/test/java/co/elastic/apm/agent/jdbc/AbstractJdbcInstrumentationTest.java]: this is now covered by tests

File: apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/StatementInstrumentation.java [hunk 4]
```diff
@@ -172,8 +258,78 @@ public static Span onBeforeExecute(@Advice.This Statement statement) throws SQLE
...
+
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class)
+        public static void onAfterExecute(@Advice.This Statement statement,
+                                          @Advice.Enter @Nullable Span span,
+                                          @Advice.Thrown Throwable t) throws SQLException {
+            if (span != null) {
+                span.getContext()
+                    .getDb()
+                    // getUpdateCount javadoc indicates that this method should be called only once
+                    // however in practice adding this extra call seem to not have noticeable side effects
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/StatementInstrumentation.java]: I have found this SO post which seems relevant: https://stackoverflow.com/questions/9984468/getresultset-should-be-called-only-once-per-result

**Pull request merged**